internal/mongoutil/mongoutil.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongoutil
+
+import (
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Options defines options types that can be merged using the functional
+// setters.
+type Options interface {
+	options.AggregateOptions | options.BucketOptions | options.BulkWriteOptions |
+		options.ClientOptions | options.ClientEncryptionOptions | options.CollectionOptions |
+		options.CountOptions | options.CreateIndexesOptions |
+		options.CreateCollectionOptions | options.CreateSearchIndexesOptions |
+		options.CreateViewOptions | options.DataKeyOptions | options.DatabaseOptions |
+		options.DefaultIndexOptions | options.DeleteOptions | options.DistinctOptions |
+		options.DropCollectionOptions | options.DropIndexesOptions |
+		options.DropSearchIndexOptions | options.EncryptOptions |
+		options.EstimatedDocumentCountOptions | options.FindOptions |
+		options.FindOneOptions | options.FindOneAndDeleteOptions |
+		options.FindOneAndReplaceOptions | options.FindOneAndUpdateOptions |
+		options.GridFSFindOptions | options.GridFSNameOptions | options.GridFSUploadOptions |
+		options.IndexOptions | options.InsertManyOptions | options.InsertOneOptions |
+		options.ListCollectionsOptions | options.ListDatabasesOptions |
+		options.ListIndexesOptions | options.ListSearchIndexesOptions |
+		options.LoggerOptions | options.RangeOptions | options.ReplaceOptions |
+		options.RewrapManyDataKeyOptions | options.RunCmdOptions |
+		options.SearchIndexesOptions | options.ServerAPIOptions | options.SessionOptions |
+		options.TimeSeriesOptions | options.TransactionOptions | options.UpdateOptions |
+		options.UpdateSearchIndexOptions | options.ChangeStreamOptions |
+		options.AutoEncryptionOptions
+}
+
+// OptionsBuilder is an interface that wraps a method to return a list of setter
+// functions for merging options for a generic options type.
+type OptionsBuilder[T Options] interface {
+	OptionsSetters() []func(*T) error
+}
+
+// NewOptionsFromBuilder will functionally merge a slice of mongo.Options in a
+// "last-one-wins" manner, where nil options are ignored.
+func NewOptionsFromBuilder[T Options](opts ...OptionsBuilder[T]) (*T, error) {
+	args := new(T)
+	for _, opt := range opts {
+		if opt == nil || reflect.ValueOf(opt).IsNil() {
```

Comment:
Can you provide documentation concerning the heap allocation for structs with use of `reflect.ValueOf`? It's not obvious to me why this function would need to allocate anything to heap in this case. For example:

```
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/internal/mongoutil
BenchmarkReflectValueOf-10       1602472               635.8 ns/op          1428 B/op         11 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/internal/mongoutil  5.912s
```

```go
func BenchmarkReflectValueOf(b *testing.B) {
	s := make([]OptionsBuilder[options.ClientOptions], 0, 1_000_000)
	for i := 0; i < 1_000_000; i++ {

		s = append(s, options.Client().ApplyURI("x").SetAppName("y").
			SetAuth(options.Credential{}).SetHosts([]string{"x", "y"}).
			SetDirect(true).SetTimeout(time.Second))
	}

	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_, _ = NewOptionsFromBuilder[options.ClientOptions](s...)
	}
}
```

Here are the results when you check for `nil` in the options setter:

```
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/internal/mongoutil
BenchmarkReflectValueOf-10       1396346               774.1 ns/op          1579 B/op         17 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/internal/mongoutil  10.938s
```

Reply:
Can you provide documentation concerning the heap allocation for structs with use of `reflect.ValueOf`? It's not obvious to me why this function would need to allocate anything to heap:

```go
func BenchmarkReflectValueOf(b *testing.B) {
	s := make([]OptionsBuilder[options.LoggerOptions], 0, 1_000_000)
	for i := 0; i < 1_000_000; i++ {
		var lo *options.LoggerOptionsBuilder

		s = append(s, lo)
	}

	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_, _ = NewOptionsFromBuilder[options.LoggerOptions](s...)
	}
}
```
```
❯ go1.18.1 test -bench=. -benchmem
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/internal/mongoutil
BenchmarkReflectValueOf-10           499           2122966 ns/op           32123 B/op          1 allocs/op
PASS
ok      go.mongodb.org/mongo-driver/internal/mongoutil  1.996s
```

Additionally, checking for nil in `OptionsSetter()` does not actually solve this issue. The `reflect.IsNil()` check is to insure that an interface containing an nil concrete type doesn't call `OptionsSetters()`. For example:

```go
var lo *options.LoggerOptionsBuilder
_, _ = NewOptionsFromBuilder[options.LoggerOptions](s...)
``` 

That is, the nil pointer dereference doesn't happen _in_ `OptionsSetters`, it happens _calling_ `OptionsSetters`. The _in_ case is covered by the first check in the condition `opt == nil`, ensuring `reflect.ValueOf` is not called.

Reply:
Also, I wouldn't expect reflect.ValueOf to functionally allocate to stack over heap for structs,  unless they are massive. For example:

```go
func BenchmarkReflectValueOf(b *testing.B) {
	b.ReportAllocs()

	example := options.Client().
		ApplyURI("x").
		SetAppName("y").
		SetAuth(options.Credential{})

	for i := 0; i < b.N; i++ {
		_ = reflect.ValueOf(example).IsNil()
	}
}
```

```
❯ go1.18.1 test -bench=. -benchmem
goos: darwin
goarch: arm64
pkg: heap
BenchmarkReflectValueOf-10      714014944                1.615 ns/op           0 B/op          0 allocs/op
PASS
ok      heap    2.033s
```

Reply:
I see the update you're referencing: https://go-review.googlesource.com/c/go/+/408826?usp=email 

I'm not sure determining how to avoid using reflection here is worth the optimization at the moment.

Reply:
Agreed, benchmarking the version with and without the `reflect.ValueOf` call shows almost no difference, even on Go 1.18. Using `reflect.ValueOf().IsNil` to detect nil interface pointers is a more robust solution than having to do that nil check in every `OptionsSetters` method, so that seems preferable. Consider this comment resolved.